ARCHITECTURE.md

```diff
@@ -0,0 +1,117 @@
+work correctly with data generated by an older plugin version within the same
+major version, degrading new features gracefully if required.
+
+
+```
+{
+    "features": [],
+    "prefixes": {},
+    "roles": {},
+    "languages": {},
+    "sourceReports": [
+        {
+            "docSetFiles": [],
+            "targetReports": [
+                {
+                    "localDocs": {},
+                    "roleDefs": {},
+                    "concepts": {},
+                    "facts": {},
+                    "rels": {},
+                }
+            ]
+        }
+    ],
+    "validation": [],
+    "filingDocuments": "foo.zip"
+}
+```
+
+A viewer can contain one or more "source reports", each producing one or more
+target report.
+
+A source report is an Inline XBRL Document Set, comprising one or more Inline
+XBRL Documents.  A target report is the XBRL data obtained from extracting a
+given target document from the source report.
+
+Target documents are not yet supported, so at present, there will always be one
+target report per source report.
+
+### prefixes
+
+`prefixes` is a map of namespace prefixes to namespace URIs.  It is shared by
+all Target Reports.
+
+### languages
+
+
+```
+"languages": {
+    "en-us": "English (US)",
+    "en": "English",
+    "fr": "French",
+},
+
+```
+
+A map of language codes to language labels.  There should be an entry here for
+every language code used by a label in a report.
+
+### sourceReports
+
+`sourceReports` is an array of objects of report data.  Prior to multi-document
+support, this property was not present, and the `docSetFiles` property, and the
+properties of the objects within `targetReports` appeared at the top-level.
+
+There is an object in this array for each separate Inline XBRL Document Set in
+the viewer.
+
+### docSetFiles
+
+`docSetFiles` is a list of files in the document set.  It is used to load other
+documents in the document set into iframes in tabs in the viewer.  For single
+file, non-stub viewers, this property is not present.
+
+### targetReports
+
+An array of objects, one for each target document obtained from the enclosing
+source report.
+
+### localDocs
+
+```
+{
+    "filename.xsd": ["schema"],
+    "inline.xhtml": ["inline"],
+}
+```
+
+`localDocs` is the set of XBRL files local to the filing (i.e. the Inline XBRL
+files, and any extension taxonomies).
+
+The property is used to populate a summary of the files in the filing in the
+document summary panel.
+
+## IDs and Viewer Unique IDs
+
+The viewer requires all iXBRL facts to have `id` attributes.  The Python plugin
+will add `id` attributes to facts if not present.  IDs are required to be unique
+within an Inline XBRL Document Set, but in order to avoid collisions between IDs
+in different Inline XBRL Document Sets when multiple source reports are
+present, the JavaScript viewer prefixes IDs with a source report index when
+loading (e.g `0-f1` is `f1` in the first report).
```

Comment:
> Then, it might help to add a sentence at the end?

Expanded description in 60a0c06f9cd59434e0e80501220874f08ad6dbb5

> If there is only one Inline IXDS, I wonder if it would be nice to continue using #f-{id} so that following a legacy deep link, clicking off, clicking back on again keeps the same URL.

Good call, although I've gone for the slightly simpler approach of always omitting the report number for report `0` (b7097cb)

> Pre-existing, it seems like the #f-{id} format could conflict with other identifiers in the HTML

Actually, I don't think this is an issue, because the 3rd party HTML is in an iframe.  Potentially you could get an unwanted scroll when loading in non-stub-viewer mode, but I think it would correct itself once the viewer loads.

Reply:
I think `<a href="#f-x">` (handled [here](https://github.com/Arelle/ixbrl-viewer/blob/90d1fc05dd21aceacc3c4fbf47d29dcb2751d729/iXBRLViewerPlugin/viewer/src/js/viewer.js#L211-L215)) is ambiguous for `<h1 id="f-x">` and `<ix:nonNumeric id="x">`?

Reply:
I don't think there's actually a conflict here.  A fragment identifier in the viewer URL links to a fact via the `ivid` data attribute, handled by `handleDeepLink`.  The code above handles links between viewer tabs via HTML `id` or  `<a name>` attributes.  There's currently no way to deep link to an HTML element in the iXBRL report from outside the viewer.

Reply:
> There's currently no way to deep link to an HTML element in the iXBRL report from outside the viewer.

It works for me (non-stub): if I put a `<div id="x">` at the bottom of the document and then open `test.htm#x`, it scrolls to the correct place.  I'm not sure how that works after `_reparentDocument`; maybe the browser doesn't process the fragment until after the index.js `onready`?

Reply:
Ah yes.  In non-stub mode the HTML elements are accessible in this way.  It's interesting that the iframe is also scrolled – it's definitely scrolling before the re-parenting happens, but then seems to somehow transfer the scroll position of the outer window onto the iframe.

I don't think that any of the options for fixing this are appealing:

* We could rewrite the HTML IDs, but that would mean modifying the source document.
* We could detect conflicts and modify the fact IDs, but that would mean that the fragments were no longer predictable from outside the app (being able to deep link to the viewer when working from xBRL-JSON, for example, is useful)

HTML linking only works for non-stub documents anyway, and the behaviour in the event of conflict is well defined: the fact link will prevail.

Reply:
> We could rewrite the HTML IDs, but that would mean modifying the source document.

This seems unworkable since it would require finding/updating `<a href="#x">` references (intra- and inter-document references).  I have also seen use-cases where people combine iXBRL viewer output + external documents (e.g., non-IXDS exhibits) and link between the two...

> We could detect conflicts and modify the fact IDs, but that would mean that the fragments were no longer predictable from outside the app (being able to deep link to the viewer when working from xBRL-JSON, for example, is useful)

Agreed.

> HTML linking only works for non-stub documents anyway

It would be convenient if that did work even for stub so that you can (e.g.) right click table of contents links and share with others.  We could presumably adjust the `handleFactDeepLink` to use the `_showDocumentAndElement` logic if it wasn't a fact identifier, right?

> and the behaviour in the event of conflict is well defined: the fact link will prevail.

It seems like we should document that behavior of fragments that XBRL generators should avoid `f-` and `f[0-9]+-` `id=""`/`<a name="">` prefixes since those are reserved for ixbrl-viewer deep links.  That seems to be trending towards [Designing HTML for Inline XBRL 1.0].

At first I thought we could use `#{number}-{id}` since XML IDs cannot start with numbers but fragments can, which would avoid conflicts with `<h1 id="x">`, but there are still conflicts with `<a name="x">`.  Anyway, this is all pre-existing, so we can always revisit it later.

Reply:
Yes, that would work.  I've raised https://github.com/Arelle/ixbrl-viewer/issues/572 to track this.